opnode/node/node.go

```diff
@@ -1,18 +1,236 @@
+	L1Num  uint64      `ask:"--l1-num" help:"Block number of L1 matching the l1-hash"`
+}
+
+func (conf *GenesisConf) GetGenesis() l2.Genesis {
+	return l2.Genesis{
+		L1: eth.BlockID{Hash: conf.L1Hash, Number: conf.L1Num},
+		// TODO: if we start from a squashed snapshot we might have a non-zero L2 genesis number
+		L2: eth.BlockID{Hash: conf.L2Hash, Number: 0},
+	}
+}
 
 type OpNodeCmd struct {
+	L1NodeAddrs   []string `ask:"--l1" help:"Addresses of L1 User JSON-RPC endpoints to use (eth namespace required)"`
+	L2EngineAddrs []string `ask:"--l2" help:"Addresses of L2 Engine JSON-RPC endpoints to use (engine and eth namespace required)"`
+
+	LogCmd `ask:".log" help:"Log configuration"`
+
+	Genesis GenesisConf `ask:".genesis" help:"Genesis anchor point"`
+
+	// during later sequencer rollup implementation:
+	// TODO: multi-addrs option (static peers)
+	// TODO: bootnodes option (bootstrap discovery of more peers)
+
+	log log.Logger
+
+	// (combined) source to fetch data from
+	l1Source eth.L1Source
+
+	// engines to keep synced
+	l2Engines []*l2.EngineDriver
+
+	l1Downloader l1.Downloader
+
+	ctx   context.Context
+	close chan chan error
+}
+
+func (c *OpNodeCmd) Default() {
+	c.L1NodeAddrs = []string{"http://127.0.0.1:8545"}
+	c.L2EngineAddrs = []string{"http://127.0.0.1:8551"}
 }
 
 func (c *OpNodeCmd) Help() string {
 	return "Run optimism node"
 }
 
 func (c *OpNodeCmd) Run(ctx context.Context, args ...string) error {
+	logger := c.LogCmd.Create()
+	c.log = logger
+	c.ctx = ctx
+
+	if c.Genesis == (GenesisConf{}) {
+		return errors.New("genesis configuration required")
+	}
+
+	l1Sources := make([]eth.L1Source, 0, len(c.L1NodeAddrs))
+	for i, addr := range c.L1NodeAddrs {
+		// L1 exec engine: read-only, to update L2 consensus with
+		l1Node, err := rpc.DialContext(ctx, addr)
+		if err != nil {
+			// HTTP or WS RPC may create a disconnected client, RPC over IPC may fail directly
+			if l1Node == nil {
+				return fmt.Errorf("failed to dial L1 address %d (%s): %v", i, addr, err)
+			}
+			c.log.Warn("failed to dial L1 address, but may connect later", "i", i, "addr", addr, "err", err)
+		}
+		// TODO: we may need to authenticate the connection with L1
+		// l1Node.SetHeader()
+		cl := ethclient.NewClient(l1Node)
+		l1Sources = append(l1Sources, cl)
+	}
+	if len(l1Sources) == 0 {
+		return fmt.Errorf("need at least one L1 source endpoint, see --l1")
+	}
+
+	// Combine L1 sources, so work can be balanced between them
+	c.l1Source = eth.NewCombinedL1Source(l1Sources)
+	l1CanonicalChain := eth.CanonicalChain(c.l1Source)
+
+	c.l1Downloader = l1.NewDownloader(c.l1Source)
+
+	for i, addr := range c.L2EngineAddrs {
+		// L2 exec engine: updated by this OpNode (L2 consensus layer node)
+		backend, err := rpc.DialContext(ctx, addr)
+		if err != nil {
+			if backend == nil {
+				return fmt.Errorf("failed to dial L2 address %d (%s): %v", i, addr, err)
+			}
+			c.log.Warn("failed to dial L2 address, but may connect later", "i", i, "addr", addr, "err", err)
```

Comment:
How does this work, do we need to call `DialContext` again or is this done implicitly in some other call?

Reply:
See comment above: "HTTP or WS RPC may create a disconnected client, RPC over IPC may fail directly"

E.g. if the HTTP URL doesn't parse, you may get a nil client + error. But if it simply fails to connect, then you get a non-nil client + error warning about the failed connection. The client will just re-attempt to connect when used later.